client/lib/signup/site-type.js

```diff
@@ -13,48 +14,42 @@ import i18n from 'i18n-calypso';
  * Some (or all) of these site types will also have landing pages.
  * A user who comes in via a landing page will not see the Site Topic dropdown.
  */
-export const allSiteTypes = [
-	{
-		type: i18n.translate( 'Blog' ),
+export const allSiteTypes = {
```

Comment:
Thanks for the inputs @michaeldcain.

> Also, to make the object simpler, I think we can use an 'id' value as both the 'value' and 'queryParam'. I don't think there's a need to have two versions (the radio form value is hidden from the user), and I think uncapitalized and statically dashed is the way to go.

Combining both fields is a good idea. I splitted it into two fields because I thought maybe we should store the unmodified site type value (i.e undashed, if that is a word) in the backend. However, on thinking through this again, I see no issues particularly in storing the uncapitalized and dashed value as the user's selection. So, the `id` field will serve the dual purpose of being the value of the site type selection, and also the slug that is used in the vertical landing pages e.g `?vertical=bar&site_type=<id>`.

> If we go the route of having our "slug" as a value inside the object (in this current case, value: 'Blog'), then I think we can keep this as an array of objects

The main reason why I went with the approach of objects ibstead of array is that the object key will remain unchanged, whereas `label` and `slug` values could change over time. So, down the line if we decide to change `Online Store` to say `Ecommerce Store`, and the corresponding slug or `id` field changes to `ecommerce-store`, the references to this site type in the code can still remain as `allSiteTypes.store`.

We're using specific references in a few places. E.g:
- In `getThemeForSiteType()`, we assign a theme depending on the site type value. 
```
if ( siteType && siteType === get( allSiteTypes, 'education.value', '' ) ) {
    return 'pub/twentyfifteen';
}
```

^^^ If `allSiteTypes` is made an array, could we write this without having to hardcode the site type value?

- In a few places([#](https://github.com/Automattic/wp-calypso/pull/28182/files#diff-c1e6c4d31974e650b129e726b3af771aR175), [#](https://github.com/Automattic/wp-calypso/blob/master/client/signup/steps/site-topic/index.jsx#L113), [#](https://github.com/Automattic/wp-calypso/pull/28604/files#diff-7949db3a74b97e79a5ff08e5b794a6d8R570)), the copy depends on the site type value, and we're hardcoding the site type value in those places. The object method helps rid us of the hardcoding and touching multiple files for segment copy changes, especially given that the segments could probably face changes in the coming weeks (we may add some, remove or edit some?).

In the above two cases, is it possible to compare the site type value without hardcoding if we use the array method? 

It's possible that I'm overthinking this, so do point me in the right direction if so 😄.

Reply:
> So, down the line if we decide to change Online Store to say Ecommerce Store, and the corresponding slug or id field changes to ecommerce-store, the references to this site type in the code can still remain as allSiteTypes.store.

I'd say in that case we would only change the name and keep ID the same, even if that would mean it says something a bit different

Reply:
@niranjan-uma-shankar: I like how you're thinking on the future code stability. I'm fine with using the object key approoach, but can we add `theme` and `designType` params to the objects and the change `getThemeForSiteType` and `getDesignTypeForSiteType` to helper functions (inside `/lib/signup/site-type.js`) that just get the values.

The following feels more like something that would be returned from the upcoming `/segments/` endpoint.

```
allSiteTypes = {
  'blog': {
    id: 'blog',
    label: i18n.translate( 'Blog' ),
    description: i18n.translate( 'Share and discuss ideas, updates, or creations.' ),
    theme: 'pub/radcliffe-2',
    designType: 'posts',
  },
  //...
};
```

Reply:
> but can we add theme and designType params to the objects and the change getThemeForSiteType and getDesignTypeForSiteType to helper functions (inside /lib/signup/site-type.js) that just get the values.

Yes indeed. We can certainly do that, was thinking the same. 👍

Reply:
After another round of discussions, we've decided to make `allSiteTypes` an array of objects as it best addresses all our needs. So, the format will  be:
```
allSiteTypes = [
  {
    id: 'blog',     
    slug: 'blog',  //used in vertical landing pages
    label: i18n.translate( 'Blog' ),
    description: i18n.translate( 'Share and discuss ideas, updates, or creations.' ),
    theme: 'pub/radcliffe-2',
    designType: 'posts',
  },
  //...
];
```

This ensures that iterating over allSiteTypes using `Array.prototype.map()` will always be [in order](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map#Description).

Reply:
This is fixed in [da96beb](https://github.com/Automattic/wp-calypso/pull/28672/commits/da96beb444e5fe822efc68f8d0047d7c17742515)